packages/zero-client/src/client/mutation-tracker.ts

```diff
@@ -64,14 +57,26 @@ export class MutationTracker {
     this.#outstandingMutations = new Map();
     this.#ephemeralIDsByMutationID = new Map();
     this.#allMutationsAppliedListeners = new Set();
-    this.#limboMutations = new Set();
     this.#largestOutstandingMutationID = 0;
     this.#currentMutationID = 0;
     this.#ackMutations = ackMutations;
   }
 
-  set clientID(clientID: string) {
+  setClientIDAndWatch(
+    clientID: string,
+    experimentalWatch: ReplicacheImpl['experimentalWatch'],
+  ) {
+    assert(this.#clientID === undefined, 'clientID already set');
     this.#clientID = clientID;
+    experimentalWatch(
```

Comment:
If I understand this correctly its essential for correctness that for a poke this watch must callback must run before zero-poke-handler calls lmidAdvanced for that poke.

I assume this order works out because ZeroPokeHandler awaits

await this.#replicachePoke(merged); 

which must not resolve until the subscribers/watchers are run

before ZeroPokeHandler calls

lmidAdvanced.

It would be good to document this ordering constraint.

With the current impl of MutationTracker I think if this ever happened out of order and a poke had responses, we would resolve the mutations for those responses with success during the lmidAdvanced call, and then drop the responses when they came in through watch.  It would be good if we could at least assert that the order is as expected.

Reply:
Yep, that's right. There are some tests that check this but adding an invariant is a good idea. We can check that the response LMID is always > than the last LMID processed by mutation tracker.